### PR TITLE
chore: Generate SBOM. Fixes #7672

### LINKED ISSUE #7672: Automate SBOM generation
# Summary

Automate [SBOM](https://www.linuxfoundation.org/blog/what-is-an-sbom/) generation and include it as part of release artifacts.
Copied from Argo CD issue #[8081](https://github.com/argoproj/argo-cd/issues/8081)

# Motivation

SBOM would be useful for end-users and partners to quickly find out which components are used by Argo Workflows and help to identify components that may be subject to security vulnerabilities, catch components with incompatible licenses etc.

# Proposal

Implement a script that would produce markdown document with the table that includes list of components, version and license. The release workflow should use the script to generated SBOM document and attach it to the Github release.

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>
Fixes #7672
Output: https://github.com/argoproj/argo-workflows/releases/tag/v0.0.0-dev-bom-2

Don't bother creating a PR until you've done this:

* [x] Run `make pre-commit -B` to fix codegen, lint, and commit message problems.

Create your PR as a draft.

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it
  does not need to pass.
* Once required tests have passed, you can make it "Ready for review".
* Say how you tested your changes. If you changed the UI, attach screenshots.

Tips:

* If changes were requested, and you've made them, then dismiss the review to get it looked at again.
* Add you organization to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md) if you like.
* You can ask for help!

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -323,13 +323,21 @@ jobs:
...
+        with:
+          go-version: "1.17"
+      - run: go install sigs.k8s.io/bom/cmd/bom@v0.2.0
+      - run: go install github.com/spdx/spdx-sbom-generator/cmd/generator@v0.0.13
+      - run: mkdir -p dist
+      - run: generator -o dist -p .
+      - run: generator -o dist -p ui
+      - run: bom generate --image quay.io/argoproj/workflow-controller:${GITHUB_REF##*/}
+      - run: bom generate --image quay.io/argoproj/argocli:${GITHUB_REF##*/}
+      - run: bom generate --image quay.io/argoproj/argoexec:${GITHUB_REF##*/}
```
Comment (leoluz) [inline .github/workflows/release.yaml:336]: You are missing the `-o` flag here to create the spdx files in the `dist` folder

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -323,13 +323,21 @@ jobs:
       - uses: actions/setup-node@v2.5.1
         with:
           node-version: "16"
+      - uses: actions/setup-go@v2
+        with:
+          go-version: "1.17"
+      - run: go install sigs.k8s.io/bom/cmd/bom@v0.2.0
+      - run: go install github.com/spdx/spdx-sbom-generator/cmd/generator@v0.0.13
```
Comment (leoluz) [inline .github/workflows/release.yaml:330]: interesting.. I tried that and it didn't work for me at first. This is better than downloading the binary. Will change argocd to use `go install` as well. 👍🏻

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -323,13 +323,21 @@ jobs:
       - uses: actions/setup-node@v2.5.1
         with:
           node-version: "16"
+      - uses: actions/setup-go@v2
+        with:
+          go-version: "1.17"
+      - run: go install sigs.k8s.io/bom/cmd/bom@v0.2.0
```
Comment (leoluz) [inline .github/workflows/release.yaml:329]: very minor but one thing I did for ArgoCD repo was parametrizing the versions as env vars to make it easier to update in future.

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -364,5 +373,6 @@ jobs:
             dist/argo-*.gz
             dist/argo-*.gz.sha256
             dist/manifests/*.yaml
+            dist/*.spdx
```
Comment (leoluz) [inline .github/workflows/release.yaml:376]: Those spdx files will be large text files. In ArgoCD and ArgoRollouts we are going to provide all spdx files packed and compacted with `tar.gz` in a file called `sbom.tar.gz`. Maybe we should maintain consistency across the projects and do the same for ArgoWorkflows?

Comment (alexec) [reply .github/workflows/release.yaml:376]: I don't think I'm worried about size, but useful to download them all in a single file. Will update

**Review by sarabala1979 — APPROVED** [review]

**Review by leoluz — APPROVED** [review]
LGTM

**Pull request merged**